### PR TITLE
[show] Add show ssd CLI

### PR DESCRIPTION
**Pull request by mudsut4ke**
**- What I did**
Add commands to display SSD information
```
  ssd_all               Execute all test iterms of SSD
  ssd_badblock          Check the later bad block status which may...
  ssd_capcity           Show ssd capcity
  ssd_firmwareinfo      Show ssd fwinfo
  ssd_health            Check the health status
  ssd_help              Show all SSD cmd
  ssd_pecycle           Show ssd P/E cycle
  ssd_remaintime        Show ssd Remaining Time
  ssd_sn                Show ssd serial number
  ssd_temperature       Read SSD temperature C
```
**- How I did it**
Add function to display ssd information to '[show/main.py](https://github.com/SW-CSV/sonic-utilities/blob/1c864f278b0e52257c90699e46e9822f45d401d2/show/main.py)' 

**- How to verify it**

1. Install smartmontool (https://github.com/Azure/sonic-buildimage/pull/2703)
2. Run show ssd command as above

**- Previous command output (if the output of a command-line utility has changed)**
```
admin@sonic:~$ show
Usage: show [OPTIONS] COMMAND [ARGS]...

  SONiC command line - 'show' command

Options:
  -?, -h, --help  Show this message and exit.

Commands:
  aaa                   Show AAA configuration
  acl                   Show ACL related information
  arp                   Show IP ARP table
  clock                 Show date and time
  ecn                   Show ECN configuration
  environment           Show environmentals (voltages, fans, temps)
  interfaces            Show details of the network interfaces
  ip                    Show IP (IPv4) commands
  ipv6                  Show IPv6 commands
  line                  Show all /dev/ttyUSB lines and their info
  lldp                  LLDP (Link Layer Discovery Protocol)...
  logging               Show system log
  mac                   Show MAC (FDB) entries
  mirror_session        Show existing everflow sessions
  mmu                   Show mmu configuration
  ndp                   Show IPv6 Neighbour table
  ntp                   Show NTP information
  pfc                   Show details of the priority-flow-control...
  platform              Show platform-specific hardware info
  priority-group        Show details of the PGs
  processes             Display process information
  queue                 Show details of the queues
  reboot-cause          Show cause of most recent reboot
  route-map             show route-map
  runningconfiguration  Show current running configuration...
  services              Show all daemon services
  startupconfiguration  Show startup configuration information
  system-memory         Show memory information
  tacacs                Show TACACS+ configuration
  techsupport           Gather information for troubleshooting
  uptime                Show system uptime
  users                 Show users
  version               Show version information
  vlan                  Show VLAN information
  warm_restart          Show warm restart configuration and state
  watermark             Show details of watermark
```

**- New command output (if the output of a command-line utility has changed)**
```
admin@sonic:~$ show
Usage: show [OPTIONS] COMMAND [ARGS]...

  SONiC command line - 'show' command

Options:
  -?, -h, --help  Show this message and exit.

Commands:
  aaa                   Show AAA configuration
  acl                   Show ACL related information
  arp                   Show IP ARP table
  clock                 Show date and time
  ecn                   Show ECN configuration
  environment           Show environmentals (voltages, fans, temps)
  interfaces            Show details of the network interfaces
  ip                    Show IP (IPv4) commands
  ipv6                  Show IPv6 commands
  line                  Show all /dev/ttyUSB lines and their info
  lldp                  LLDP (Link Layer Discovery Protocol)...
  logging               Show system log
  mac                   Show MAC (FDB) entries
  mirror_session        Show existing everflow sessions
  mmu                   Show mmu configuration
  ndp                   Show IPv6 Neighbour table
  ntp                   Show NTP information
  pcie_checkid          Chedk Vender ID and Devie ID Check
  pcie_help             Show all SSD cmd
  pcie_lnkspeed         Chedk the link speed
  pfc                   Show details of the priority-flow-control...
  platform              Show platform-specific hardware info
  priority-group        Show details of the PGs
  processes             Display process information
  queue                 Show details of the queues
  reboot-cause          Show cause of most recent reboot
  route-map             show route-map
  runningconfiguration  Show current running configuration...
  services              Show all daemon services
  ssd_all               Execute all test iterms of SSD
  ssd_badblock          Check the later bad block status which may...
  ssd_capcity           Show ssd capcity
  ssd_firmwareinfo      Show ssd fwinfo
  ssd_health            Check the health status
  ssd_help              Show all SSD cmd
  ssd_pecycle           Show ssd P/E cycle
  ssd_remaintime        Show ssd Remaining Time
  ssd_sn                Show ssd serial number
  ssd_temperature       Read SSD temperature C
  startupconfiguration  Show startup configuration information
  system-memory         Show memory information
  tacacs                Show TACACS+ configuration
  techsupport           Gather information for troubleshooting
  uptime                Show system uptime
  users                 Show users
  version               Show version information
  vlan                  Show VLAN information
  warm_restart          Show warm restart configuration and state
  watermark             Show details of watermark

```

### CODE REVIEW

**Comment by mudsut4ke** [comment]
@sandycelestica

File: show/main.py [hunk 1]
```diff
@@ -1804,6 +1804,716 @@ def tablelize(keys, data, enable_table_keys, prefix):
...
+# 'pcie checkid' command ####
+#
+
+@pcie.command()
+def checkid():
+    """show Vender ID and Devie ID Check"""
+
+    checkin = 1
+    testname = "Vender ID and Devie ID Check Test"
+    command = "sudo lspci -n | grep b960 "
```
[the remark this answers is not in the document]
Comment (sandycelestica) [reply show/main.py]: b960 is the switch chip(BCM56960) of Seastone. Here we just check switch chip info.

**Comment by jleveque** [comment]
@mudsut4ke, @sandycelestica: Are you aware of this in-progress design for SSD health: https://github.com/Azure/SONiC/wiki/SSD-Health-design-(Draft)?

File: show/main.py [hunk 2]
```diff
@@ -1804,6 +1804,716 @@ def tablelize(keys, data, enable_table_keys, prefix):
...
+            if "Device Model" in line:
+                checkin = 1
+                if ("3ME3" in line) or ("3ME4" in line):
+                    cycle = 3000
+                elif ("3IE3" in line):
+                    cycle = 20000
+                else:
+                    checkin = 0
+                    click.echo(line.strip())
+                    click.echo("Device Model Not Match 3ME3 3ME4 or 3IE3")
```
[the remark this answers is not in the document]
Comment (sandycelestica) [reply show/main.py]: Thanks for your suggest. Currently this SSD diag is just for Seastone. So not support other types of SSD right now. And we are planning to extend it.

File: show/main.py [hunk 3]
```diff
@@ -1804,6 +1804,716 @@ def tablelize(keys, data, enable_table_keys, prefix):
...
+        return
+    else:
+        for line in output:
+            if "Power_On_Hours" in line:
+                rawval = line.split()[-1]
+                poweron = int(rawval)
+                checkin = 1
+
+    if(checkin == 1):
+        remainingtime = poweron * 100 / 7.62 - poweron
```
[the remark this answers is not in the document]
Comment (sandycelestica) [reply show/main.py]: There is a formula to calculate this value:
ID Hex Attribute
----------------------
09 (09) Power on Hours
To estimate remaining hours of use: ID# 9 (100/7.26) – 1 (x ID# 9)


In our code,
if "Power_On_Hours" in line:
                rawval = line.split()[-1]
                poweron = int(rawval)
                checkin = 1

we get the poweron first and then calculate the remaining time follow the formula.

**Comment by sandycelestica** [comment]
> show utility is mostly a wrapper
> I suggest to have a separate utility (like sfputil, psuutil, intfutil, etc) and call it from show

From the beginning we discussed with Guo Han. He suggested to modify the main.py in show. So we merged it into show.

File: show/main.py [hunk 4]
```diff
@@ -1804,6 +1804,716 @@ def tablelize(keys, data, enable_table_keys, prefix):
...
     state_db.close(state_db.STATE_DB)
+#
+#echo empty line
+#
+def echo_empty_line():
+    click.echo("\n")
+
+
+#
+# function name : print_test_title
```
[the remark this answers is not in the document]
Comment (sandycelestica) [reply show/main.py]: I'll update this part later.If you have the detail request of the content.Please let me know.I plan to display like this:"Show SSD Firmware info"

**Comment by mudsut4ke** [comment]
Add @grantshao

**Comment by mudsut4ke** [comment]
> > show utility is mostly a wrapper
> > I suggest to have a separate utility (like sfputil, psuutil, intfutil, etc) and call it from show
> 
> From the beginning we discussed with Guo Han. He suggested to modify the main.py in show. So we merged it into show.

 @lguohan, Do you have a comment ?

File: show/main.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (andriymoroz-mlnx) [reply show/main.py]: I just don't understand why remaining time is calculated as poweron*12
why remaining is always greater?

File: show/main.py [hunk 5]
```diff
@@ -1804,6 +1804,716 @@ def tablelize(keys, data, enable_table_keys, prefix):
...
+
+
+#
+# 'ssd all' command 
+# execute all test iterms of SSD
+#
+
+@ssd.command()
+@click.argument("device")
+def all(device):
```
Comment (andriymoroz-mlnx) [inline show/main.py]: all() can call other functions and not duplicate the code...

File: show/main.py [hunk 6]
```diff
@@ -1804,6 +1804,716 @@ def tablelize(keys, data, enable_table_keys, prefix):
...
+    click.echo("{0:30s}{1:<40}".format("    ssd pecycle /dev/xxx     ",	"show SSD P/E cycle"))
+    click.echo("{0:30s}{1:<40}".format("    ssd health /dev/xxx      ",	"show SSD health status which should be > 95% after MFG test"))
+    click.echo("{0:30s}{1:<40}".format("    ssd badblock /dev/xxx    ",	"show SSD bad block status which may created during test"))
+    click.echo("{0:30s}{1:<40}".format("    ssd temperature /dev/xxx ",	"read SSD temperature; range 0~70 C"))
+    click.echo("{0:30s}{1:<40}".format("    ssd all /dev/xxx         ",	"show all test iterms of SSD "))
+    click.echo("{0:30s}{1:<40}".format("    ssd help /dev/xxx        ",	"list help menu \n"))
+
+
+#
+# check_pcie_speed
```
Comment (andriymoroz-mlnx) [inline show/main.py]: It would be better to have pcie in a separate PR
or at least fix the subject...

File: show/main.py (hunk 5 above)
Comment (sandycelestica) [reply show/main.py]: OK,I'll update it as your suggestion.Thanks

**Comment by liat-grozovik** [comment]
We had an SSD HLD in the last community meeting. I think merging this CLI now will contradict the design held. The design is also referred earlier on this PR discussion.
My suggestion is to close the PR and to wait for the SSD implementation by Andiry Moroz.

@andriymoroz-mlnx could you please share the design document in https://github.com/Azure/SONiC/tree/master/doc as PRfollowing the review comments?

**Comment by andriymoroz-mlnx** [comment]
> @andriymoroz-mlnx could you please share the design document in https://github.com/Azure/SONiC/tree/master/doc as PRfollowing the review comments?

Created https://github.com/Azure/SONiC/pull/378/files
